uwblueprint/extend-a-family
# Help request notifications/fixed

## Notion ticket link

[Ticket Name](https://www.notion.so/uwblueprintexecs/7e51fde5f22d448b8226f2cfa4dd93d5?v=b2dc0dba06574651a8b6063bdd0d74e0&p=f9fd876634c4411691742f6e7790da71&pm=s)

![large-pr](https://github.com/user-attachments/assets/4280623a-3973-4128-bdbc-3ba44806f628)



## Implementation description
* Implemented notifications and help request table
* Notifications:
  1) Allows all users (currently only notifies facilitators) to receive real-time notifications
  2) Displays a notification bell that signifies how many unseen notifications there are and goes away after closing the bell
  3) Notification menu is paginated to load 10 at a time (doesn't actually but it should)
* Help requests:
  1) Allow learners to create help requests that notifies their facilitator
  2) Displays everything in a table sorted by time created and provides access to the corresponding module/unit/page
  3) Each help request can be marked as completed/uncompleted and clicking on id brings to the individual page




## Steps to test

I might add a video about doing this cuz these steps are too hard for me to understand too ngl

1. You will need a learner account and a facilitator user where they reference each other in the respective field
2. Open two browser instances and login both accounts. 
3. On the facilitator side, go to /help-requests (you actually need to do this step before step 4)
4. On the learner side go to /ask-for-help and click on the modal and type a request
5. On facilitator side (for steps 5-10), you should see that you get a notification and click on the bell icon will display the message
6. Upon closing the bell, the red circle with the number should go away
7. Click on the notification in the menu and it should take you to /help-requests/:id where id is the help request id. You should see the message and the id of the request
8. Change the id to one that doesn't exist in the database and it should say help request doesn't exist or something
9. Go back to /help-requests by clicking on the link in the top left and toggle the help request as completed. Refresh the page to see if it persists.
10. Make sure that click on the help request id brings you to the same page as the notification too



## What should reviewers focus on?
* ngl this is too long so whoever is reviewing this have fun. Most of the bad code is on the backend though
* I am prolly very inconsistent so like tell me where i should change variable names and stuff

* In notification router, there is a /test route. That shouldn't exist but it's only there since I haven't figured out how to fix an error related to importing mongoose models in the correct order. Dont worry abt testing that. 

![2l7u3sqgemo21](https://github.com/user-attachments/assets/45c6072e-ca7c-4028-8040-d7f281a037bd)



## Checklist
